### ISSUE TITLE
404-handling overlooks... missing files

### ISSUE THREAD

**Issue by adamcameron** [cfwheels/cfwheels#1065]
There's a slight bug here in the [`$abortInvalidRequest` handler in `internal.cfm`](https://github.com/cfwheels/cfwheels/blob/master/wheels/global/internal.cfm#L437):

```
public void function $abortInvalidRequest() {
	local.applicationPath = Replace(
		GetCurrentTemplatePath(),
		"\",
		"/",
		"all"
	);
	local.callingPath = Replace(GetBaseTemplatePath(), "\", "/", "all");
	if (
		ListLen(local.callingPath, "/") > ListLen(local.applicationPath, "/")
		|| GetFileFromPath(local.callingPath) == "root.cfm"
	) {
		if (StructKeyExists(application, "wheels")) {
			if (StructKeyExists(application.wheels, "showErrorInformation") && !application.wheels.showErrorInformation) {
				$header(statusCode = 404, statustext = "Not Found");
			}
			if (StructKeyExists(application.wheels, "eventPath")) {
				$includeAndOutput(template = "#application.wheels.eventPath#/onmissingtemplate.cfm");
			}
		}
		abort;
	}
}
```

It's possible to fall through the logic and hit that `abort`, in which case one just gets a largely empty response (some bled whitespace mismanagement is all), and a **200-OK** status. This is easy to replicate, just browse to `/any/path/does/not/matter/as/long/as/it/does/not/exist/index.cfm`. Doing this yer - possibly _by definition_ - not going to be in the `application.wheels` context, so you will never set the 404 header.

I can't help but think that in situations where `application.wheels` actually exists, it means you intrinsically must be on a good URL? Otherwise how would the application be identified to have an application scope? Not so sure about that, but perhaps more logic here needs to be checked.

Also it might also be able to fall right through to the bottom of the method entirely, in which case... there's no handling at all. I have not tried to induce this.

I'm pretty sure you want to have this just before the end of the method, outside any of the flow control statements:

```
$header(statusCode = 404, statustext = "Not Found");
abort;
```

As an aside: this code is triggered from `onApplicationStart`. I suspect it ought to be being triggered in `onRequestStart`, given it's related to requests, not the application as a whole..? Different issue though.

**Comment by adamcameron** [comment]
I encountered another "unexpected" behaviour of this last night, and discussed with @neokoenig. In DEV mode, the "404" page doesn't actually return a 404; to returns a 200. Now it makes sense for the 404 message to be more helpful to a dev "you've forgotten to create your view or your route or whatevs, and here's some debug bumpf too", but the returned status should still be a 404. Is there a valid reason I'm missing as to why it isn't?

**Comment by neokoenig** [comment]
> ... but the returned status should still be a 404. Is there a valid reason I'm missing as to why it isn't?

I agree this should return a 404 in Dev mode; Can't see a reason why it shouldn't.

**Comment by perdjurner** [comment]
The original reason to not return 404 in development mode was most likely to make sure that the browser did not override the helpful Wheels message and show its own 404 page. Not sure how browsers handle this nowadays though.

**Comment by perdjurner** [comment]
It's possible that it's just an empty abort there because we can't be sure the $header function even exists at that point in the code.

Another reason could be related to my previous comment. If there is no wheels struct we don't know if they're in development mode and can't decide between 404 or no 404 based on that so we just default to 200.

**Comment by neokoenig** [comment]
> The original reason to not return 404 in development mode was most likely to make sure that the browser did not override the helpful Wheels message and show its own 404 page. Not sure how browsers handle this nowadays though.

If you got to https://www.google.com/890asdasd%20askjl for instance, that returns a 404 header status: the only other time I can think that a 404 would cause odd browser behaviour is a browser extension (where it tries to "search something useful" for you when you hit a 404) or a DNS portal, i.e Internet provider hi-jacking the request; custom 404 handlers are ubiquitous, and display wesbite generated content, so I don't think is an issue anymore.

**Comment by neokoenig** [comment]
Fixed in #1067
